jsonschema/src/keywords/contains.rs

```diff
@@ -422,38 +423,55 @@ pub(crate) fn compile<'a>(
         }
         #[cfg(feature = "draft201909")]
         Draft::Draft201909 => {
-            if let Some(min_contains) = parent.get("minContains") {
-                if let Some(min_contains) = min_contains.as_u64() {
-                    if let Some(max_contains) = parent.get("maxContains") {
-                        if let Some(max_contains) = max_contains.as_u64() {
-                            Some(MinMaxContainsValidator::compile(
-                                schema,
-                                context,
-                                min_contains,
-                                max_contains,
-                            ))
-                        } else {
-                            Some(Err(ValidationError::schema(schema)))
-                        }
-                    } else {
-                        Some(MinContainsValidator::compile(schema, context, min_contains))
-                    }
-                } else {
-                    Some(Err(ValidationError::schema(schema)))
-                }
-            } else if let Some(max_contains) = parent.get("maxContains") {
-                if let Some(max_contains) = max_contains.as_u64() {
-                    Some(MaxContainsValidator::compile(schema, context, max_contains))
-                } else {
-                    Some(Err(ValidationError::schema(schema)))
+            let min_contains = match map_get_u64(parent, context, "minContains").transpose() {
+                Ok(n) => n,
+                Err(err) => return Some(Err(err)),
+            };
+            let max_contains = match map_get_u64(parent, context, "maxContains").transpose() {
+                Ok(n) => n,
+                Err(err) => return Some(Err(err)),
+            };
+
+            match (min_contains, max_contains) {
+                (Some(min), Some(max)) => {
+                    Some(MinMaxContainsValidator::compile(schema, context, min, max))
                 }
-            } else {
-                Some(ContainsValidator::compile(schema, context))
+                (Some(min), None) => Some(MinContainsValidator::compile(schema, context, min)),
+                (None, Some(max)) => Some(MaxContainsValidator::compile(schema, context, max)),
+                (None, None) => Some(ContainsValidator::compile(schema, context)),
             }
         }
     }
 }
 
+#[cfg(feature = "draft201909")]
+#[inline]
+pub(crate) fn map_get_u64<'a>(
+    m: &'a Map<String, Value>,
+    context: &CompilationContext,
+    type_name: &str,
+) -> Option<Result<u64, ValidationError<'a>>> {
+    let value = m.get(type_name)?;
+    if matches!(value.as_i64(), Some(n) if n < 0) {
```

Comment:
I really like that you factored out this logic into a separate function :) Maybe it could also be moved to the `helpers.rs` file to reuse it in other keywords.

Though, there are two calls to `as_<type>` functions unconditionally, the previous approach would have only one in the valid input scenario.

It would be nice to move this block to the `None` branch of the `match` statement below to keep the happy path simpler. Then, the `n < 0` check will not be needed as well - it could be as simple as `value.is_i64()` (as all non-negative values will be covered by `as_u64()`)

Reply:
Actually you're right about 2 calls. :100: